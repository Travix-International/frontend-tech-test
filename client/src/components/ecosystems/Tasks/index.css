.ellipsis {
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap; }

.tasks {
  align-items: center;
  display: flex;
  flex: 1;
  flex-direction: column;
  justify-content: center; }

.tasks-title {
  color: rgba(255, 255, 255, 0.7);
  font-size: 42px;
  font-weight: 200; }
