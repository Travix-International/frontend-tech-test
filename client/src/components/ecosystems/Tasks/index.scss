@import 'src/styles/base/colors.scss';
@import 'src/styles/base/container.scss';
@import 'src/styles/base/fonts.scss';

.tasks {
  align-items: center;
  display: flex;
  flex: 1;
  flex-direction: column;
  justify-content: center;
}

.tasks-title {
  color: $white;
  font-size: $font-big;
  font-weight: $font-thin;
}
