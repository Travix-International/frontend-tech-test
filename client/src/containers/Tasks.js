import React, { useEffect } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import Box from '@material-ui/core/Box';
import Paper from '@material-ui/core/Paper';
import List from '@material-ui/core/List';
import ListItem from '@material-ui/core/ListItem';
import ListItemText from '@material-ui/core/ListItemText';
import ListItemIcon from '@material-ui/core/ListItemIcon';
import ListItemSecondaryAction from '@material-ui/core/ListItemSecondaryAction';
import IconButton from '@material-ui/core/IconButton';
import Checkbox from '@material-ui/core/Checkbox';
import DeleteRoundedIcon from '@material-ui/icons/DeleteRounded';
import CreateRoundedIcon from '@material-ui/icons/CreateRounded';

import { taskApi } from '../api/api';

const Tasks = ({ tasks }) => {
  useEffect(() => {
    taskApi.getTasks();
  }, []);

  return (
    <Box component={Paper} height="100vh">
      <List>
        {
          tasks.map(task => (
            <ListItem key={task.id} button>
              <ListItemIcon>
                <Checkbox
                  checked={task.completed}
                  edge="start"
                  onClick={() => taskApi.updateTask({ ...task, completed: !task.completed })}
                  tabIndex={-1}
                />
              </ListItemIcon>
              <ListItemText primary={task.title} secondary={task.description} />
              <ListItemSecondaryAction>
                <IconButton aria-label="Delete Task" edge="end" onClick={() => taskApi.deleteTask(task.id)}>
                  <DeleteRoundedIcon />
                </IconButton>
                <IconButton aria-label="Update Task" edge="end">
                  <CreateRoundedIcon />
                </IconButton>
              </ListItemSecondaryAction>
            </ListItem>
          ))
        }
      </List>
    </Box>
  );
};

Tasks.propTypes = {
  tasks: PropTypes.array.isRequired,
};
const mapStateToProps = state => ({ tasks: state.tasks });

export default connect(mapStateToProps, null)(Tasks);
