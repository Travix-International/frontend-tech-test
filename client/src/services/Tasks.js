import request from 'superagent';

export default class Tasks {
  constructor(API_HOST) {
    this.API_HOST = API_HOST || '';
  }

  fetchTasks() {
    return new Promise((resolve, reject) => {
      request
        .get(this.API_HOST + '/api/tasks')
        .then((response) => {
          resolve({ tasks: response.body.tasks });
        })
        .catch((err) => {
          reject('We are unable to fetch tasks on this moment.');
        });
    });
  }
}
