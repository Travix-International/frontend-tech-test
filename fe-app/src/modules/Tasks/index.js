import AddTasks from './AddTasks/AddTasksContainer';
import EditTasks from './EditTasks/EditTasksContainer';
import ListTasks from './ListTasks/ListTasksContainer';

export {
  AddTasks,
  EditTasks,
  ListTasks,
};
