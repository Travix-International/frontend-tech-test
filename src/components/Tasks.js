import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { fetchTasks, clearErrors } from '../redux/actions';
import Task from './Task';

class Tasks extends PureComponent {
  constructor(...args) {
    super(...args);

    this.onDismissError = this.onDismissError.bind(this);
  }

  componentDidMount() {
    this.props.dispatch(fetchTasks());
  }

  onDismissError() {
    this.props.dispatch(clearErrors());
  }

  render() {
    const { isLoading, tasks, drafts, filterText, error } = this.props;
    const taskCount = tasks.length;
    const draftCount = drafts.length;

    return (
      <div className="tasks">
        {isLoading && <p className="text-center">Retrieving your tasks...</p>}
        {error &&
          <div className="alert alert-danger" role="alert">
            <button aria-label="Close" className="close" onClick={this.onDismissError} type="button">
              <span aria-hidden="true">&times;</span>
            </button>
            {error}
          </div>
        }
        {filterText !== '' &&
          <div className="alert alert-light" role="alert">
            {taskCount === 0 ? 'No' : taskCount} results for <strong>{filterText}</strong>
          </div>
        }
        {filterText === '' && taskCount === 0 && draftCount === 0 &&
          <p className="text-center">No tasks added.</p>
        }
        <div className="card-columns">
          {tasks.map(task => <Task key={task.id} {...task} />)}
          {drafts.map((draft, i) => <Task key={i} {...draft} />)}
        </div>
      </div>
    );
  }
}

Tasks.propTypes = {
  dispatch: PropTypes.func.isRequired,
  isLoading: PropTypes.bool,
  tasks: PropTypes.array,
  drafts: PropTypes.array,
  filterText: PropTypes.string,
  error: PropTypes.string
};

const mapStateToProps = (state) => {
  const { isLoading, tasks, drafts, filterText, error } = state;

  const normalizedFilterText = filterText.toLowerCase().trim();
  const filteredTasks = filterText === '' ? tasks : tasks.filter(({ title }) =>
    title.toLowerCase().indexOf(normalizedFilterText) !== -1
  );

  return { isLoading, tasks: filteredTasks, drafts, filterText, error };
};

export default connect(mapStateToProps)(Tasks);
