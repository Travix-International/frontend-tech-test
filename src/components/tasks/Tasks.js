import React from 'react'
import PropTypes from 'prop-types'
import Spinner from 'components/shared/spinner'
import Masonry from 'react-masonry-component'
import Task from './task/Task'

const masonryOptions = {
  transitionDuration: 0,
  horizontalOrder: true
}

function Tasks({ tasks, filter, checkIsPossibleToAddSubTask }) {
  if (!tasks) {
    return <Spinner />
  }

  return (
    <Masonry
      className={'tasks-list'}
      elementType={'div'}
      options={masonryOptions}
    >
      {filter(tasks).map((x, i) => (<Task
        key={x.id}
        task={x}
        isPossibleToAddSubTask={checkIsPossibleToAddSubTask(x)}
      />))}
    </Masonry>
  )
}

Tasks.propTypes = {
  tasks: PropTypes.array,
  filter: PropTypes.func,
  checkIsPossibleToAddSubTask: PropTypes.func.isRequired
}

Tasks.defaultProps = {
  filter: items => items
}

export default Tasks
