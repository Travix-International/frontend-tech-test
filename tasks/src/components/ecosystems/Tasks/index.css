.tasks {
  align-items: center;
  display: flex;
  flex: 1;
  flex-direction: column;
  justify-content: center; }

.tasks-title {
  color: #fff;
  font-size: 42px;
  font-weight: 200; }
