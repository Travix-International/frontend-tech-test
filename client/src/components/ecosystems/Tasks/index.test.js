import React from 'react'
import ShallowRenderer from 'react-test-renderer/shallow'
import Tasks from './index'

describe('Tasks', () => {
  it('should render correctly', () => {
    const renderer = new ShallowRenderer()
    renderer.render(
      <Tasks />
    )
    const result = renderer.getRenderOutput()

    expect(result).toMatchSnapshot()
  })
})
