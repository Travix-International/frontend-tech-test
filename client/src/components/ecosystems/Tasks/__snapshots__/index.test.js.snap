// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tasks should render correctly 1`] = `
<div
  className="tasks"
>
  <h1
    className="tasks-title"
  >
    Tasks
  </h1>
  <Connect(TaskCreator) />
  <Connect(TaskList) />
</div>
`;
